### PR TITLE
Workflows: add documentation

### PR DESCRIPTION
**Pull request by vladsavelyev**
Add some documentation for the workflows module, along with motivation. This doc would be linked in the `production_pipelines` README.

### CODE REVIEW

File: cpg_utils/workflows/README.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
+# Hail Batch Workflows
+
+## Motivation
+
+[Hail Batch](https://hail.is/#hail-batch) is a powerful system for workflow programming and execution using cloud compute and storage. In CPG, we manage our own Hail Batch deployment, along with the [analysis-runner](https://github.com/populationgenomics/analysis-runner) tool that provides an approachable and secure way to submit analysis scripts to Hail Batch.
+
+In Hail, the workflows can be described using Python, and be built to be parts of more complex applications. E.g. the analysis-runner uses Hail Batch to drive itself, and the genomic variation analysis tool  called Hail Query can use Hail Batch as a backend. All that makes Hail Batch a natural choice to design genomics workflows.
```
Comment (lgruen) [inline cpg_utils/workflows/README.md]:
```suggestion
In Hail, the workflows can be described using Python, and be built to be parts of more complex applications. E.g. the analysis-runner uses Hail Batch to drive itself, and the genomic variation analysis tool called Hail Query can use Hail Batch as a backend. All that makes Hail Batch a natural choice to design genomics workflows.
```

File: cpg_utils/workflows/README.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+Classical genomics best-practices workflow implementations like [WARP](https://github.com/broadinstitute/warp) and [GATK-SV](https://github.com/broadinstitute/gatk-sv) use WDL to be run with [Cromwell](https://github.com/broadinstitute/cromwell) or Terra. In order to integrate WDL workflows into a Hail Batch infrastructure, there are few possible approaches.
```
Comment (lgruen) [inline cpg_utils/workflows/README.md]:
```suggestion
Classical genomics best-practices workflow implementations like [WARP](https://github.com/broadinstitute/warp) and [GATK-SV](https://github.com/broadinstitute/gatk-sv) use WDL to be run with [Cromwell](https://github.com/broadinstitute/cromwell) or Terra. In order to integrate WDL workflows into a Hail Batch infrastructure, there are a few possible approaches.
```

File: cpg_utils/workflows/README.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+1. Start a Cromwell instance as part of Hail Batch job script, and submit a sub-workflow inside that job,
+2. Maintain a managed Cromwell instance, and wrap only workflow submissions in a Hail Batch job,
+3. Fully re-engineer WDL workflows in Python and Hail Batch.
+
+For the GATK-SV workflow, we went with the approach #2. The workflow is quite complex, it uses many layers of branching nested sub-workflows, and it implements even minor file processing commands as separate WDL tasks. We also do not intend to adjust this workflow much as of now. In order to run GATK-SV, we support a managed Cromwell instance along with a [Python helper](https://github.com/populationgenomics/analysis-runner/blob/main/analysis_runner/cromwell.py#L230) that warps interaction with Cromwell using Hail Batch jobs, and we utilize Cromwell API along with [Swagger](https://cromwell.populationgenomics.org.au/swagger/index.html?url=/swagger/cromwell.yaml#/Workflows/metadata) to explore Cromwell logs.
```
Comment (lgruen) [inline cpg_utils/workflows/README.md]:
```suggestion
For the GATK-SV workflow, we went with the approach #2. The workflow is quite complex, it uses many layers of branching nested sub-workflows, and it implements even minor file processing commands as separate WDL tasks. We also do not intend to adjust this workflow much as of now. In order to run GATK-SV, we support a managed Cromwell instance along with a [Python helper](https://github.com/populationgenomics/analysis-runner/blob/main/analysis_runner/cromwell.py#L230) that wraps interaction with Cromwell using Hail Batch jobs, and we utilize Cromwell API along with [Swagger](https://cromwell.populationgenomics.org.au/swagger/index.html?url=/swagger/cromwell.yaml#/Workflows/metadata) to explore Cromwell logs.
```

File: cpg_utils/workflows/README.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
...ating workflows scratch, we implemented some abstractions in this module, namely a workflow "stage" with expected inputs, outputs, and dependencies, and "stage" target ("sample", "dataset", "cohort"). Unlike WDL's concept of nested sub-workflows, Hail Batch is intrinsically lower-level, providing only one layer of jobs, with each job corresponding to on bash script that is run on one cloud VM. That doesn't always match a higher-level task in the application domain. For example, a workflow stage called "genotype" can include multiple jobs to generate calling intervals, partition inputs, run the genotyping tool in parallel instances, gather results, perform post-processing and filtering. That "stage" would sit between an "alignment" stage and a "joint calling" stage, and a user might need to start a workflow from a specific stage, reusing pre-existing outputs; or run the pipeline only up to a specific stage. So it's helpful to abstract a set of jobs related in a domain-specific way as a stage, with an ability to restart/skip/select stages. It also turns out that GATK-SV facilitates from these abstractions too, as it consists of multiple disconnected sub-workflows of different levels.
```
Comment (lgruen) [inline cpg_utils/workflows/README.md]: > providing only one layer of jobs

That's not entirely true anymore with open batches (just released as part of 0.2.101) 😄 .

File: cpg_utils/workflows/README.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+![uml](docs/classes.png)
+
+This library provides an abstraction interface on top of Hail Batch jobs, namely a `Stage` abstract class, parametrised by the `Target` abstract class, representing what target this stage works with. 
+
+For example, a stage that performs read alignment to produce a CRAM file, would be derived from `SampleStage`, and be parametrised by `Sample`, and a stage that performs joint-calling would be derived from `CohortStage`, parametrised by `Cohort`.
+
+Each `Stage` final implementation must declare paths to the outputs it would write in the `expected_outputs()` method. It must also define how jobs are added into Hail Batch using the `queue_jobs()` method.
+
+A stage can also be wrapped
```
Comment (lgruen) [inline cpg_utils/workflows/README.md]: Did you mean to add more content here?

File: cpg_utils/workflows/README.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+## Stages
+
+![uml](docs/classes.png)
```
Comment (lgruen) [inline cpg_utils/workflows/README.md:23]: Another awesome diagram! Would be great to add an editable source of this as well.

E.g. for our storage policies graphs we use a simple `dot` script: https://github.com/populationgenomics/team-docs/tree/main/storage_policies/figures

File: .markdownlint.json [hunk 7]
```diff
@@ -8,6 +8,7 @@
     "allowed_elements": ["details", "img"]
   },
   "ul-indent": false,          // To allow indenting the whole list to distinguish it visually
-  "no-multiple-blanks": false  // To allow multiple blank lines between a header and
+  "no-multiple-blanks": false, // To allow multiple blank lines between a header and
                                // the previous paragraph
+  "no-trailing-spaces": false
```
Comment (lgruen) [inline .markdownlint.json:13]: Don't IDEs typically remove these automatically?

**Review by lgruen — APPROVED** [review]

**Comment by lgruen** [comment]
This is really helpful! Thanks very much. 🌟

File: cpg_utils/workflows/README.md (hunk 6 above)
Comment (vladsavelyev) [reply cpg_utils/workflows/README.md:23]: It's designed in a Miro board, I can just add a link to the board

**Pull request merged**